### ISSUE TITLE
Can I overwrite shuffle=False when using rect=True if all my images have the same size ?

### ISSUE THREAD

**Issue by Loic-76** [ultralytics/ultralytics#20605]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I wonder if I can without a risk overwrite shuffle=False with rect=True as all my images have the same shape and aspect ratio.
I feel like I lose a lot of training time with padding. What do you guys think ?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Loic-76, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for guidance—especially the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, which address many common training questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further.

If this is a custom training ❓ Question (like yours), please share as much detail as possible, such as dataset image samples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

You’re also invited to join the Ultralytics community! For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧. For deeper discussions, try [Discourse](https://community.ultralytics.com/), or check out our [Subreddit](https://reddit.com/r/Ultralytics) for community knowledge sharing.

## Upgrade

To ensure you aren't encountering an issue that has been resolved in the latest release, upgrade the `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

You can use YOLO in any of these verified, up-to-date environments (with all dependencies such as [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response 🤖. An Ultralytics engineer will review your question and assist you soon!

**Comment by Y-T-G** [comment]
If you have `mosaic` enabled, then the padding isn't just an empty area. It's portions of other images. So it's not really a wasted space.

**Comment by igs-raiderler** [comment]
This is a well-structured codebase. I appreciate the clean separation of concerns.

**Comment by robinson-srk** [comment]
> If you have `mosaic` enabled, then the padding isn't just an empty area. It's portions of other images. So it's not really a wasted space.

Indeed, did not think about this one ! Probably a cleaner solution indeed as it solve the "wasted space issue" and probably over-fitting, gonna try this one to see how the padding is done with mosaic enabled.

**Comment by Loic-76** [comment]
> If you have `mosaic` enabled, then the padding isn't just an empty area. It's portions of other images. So it's not really a wasted space.

Hello and thank you for your reply, I tested it and I want to add a little something.

As the center of the generated mosaic is determined using random values (and can either be inside the image or outside of it), the imgsz parameter must be adapted so there is no padding.
For example with imgsz=2000 on 1000x2000 images, there was still some padding added.

**Comment by glenn-jocher** [comment]
You're right about the mosaic center placement affecting padding. Since all your images have the same shape and aspect ratio, in theory you could enable shuffling with rect mode without affecting padding optimization. The `rect=True` setting primarily helps when you have varying aspect ratios by sorting images to minimize padding in batches.

When images all have identical dimensions, that sorting becomes less critical. You could modify the code to allow shuffling in this specific case, but be aware this isn't a standard configuration. If you implement this change, I recommend testing the performance impact thoroughly.

For your padding issue with 1000x2000 images, setting `imgsz=[1000,2000]` (specifying both height and width) rather than just `imgsz=2000` might help ensure the correct aspect ratio is maintained throughout the pipeline.

**Comment by UltralyticsAssistant** [comment]
A potential fix is now merged in [PR #23419: “Disable shuffle only if batch shapes mismatch with `rect=True`”](https://github.com/ultralytics/ultralytics/pull/23419).

### What changed (and why it may resolve this)
The DataLoader shuffle logic for YOLO detection training was previously overly strict: when `rect=True` and `mode=="train"`, shuffling was always disabled and a warning was shown.  
With this update, shuffling is **only** disabled when it’s truly incompatible—specifically when `rect=True` **and** the dataset uses **multiple batch shapes** (i.e., `batch_shapes` vary across batches). This should:
- prevent shuffle from being turned off unnecessarily,
- reduce warning noise,
- improve training behavior/UX in `rect=True` setups where a single consistent batch shape is used.

### Credit
Merged thanks to @Y-T-G, with contributions from @glenn-jocher.

### How to test the fix
No specific version number was referenced in the PR, so the easiest ways to verify are:

- Install from the latest `main`:
  - `pip install git+https://github.com/ultralytics/ultralytics.git@main`

- Or upgrade normally and wait for the next official PyPI release:
  - `pip install -U ultralytics`

After updating, re-run the same training command that triggered the issue and confirm whether:
- shuffling remains enabled when `rect=True` and batch shapes are consistent, and/or
- the “incompatible with shuffle” warning only appears when batch shapes actually vary.

### Feedback requested
Confirmation on whether this change resolves the issue in the original reproduction would be really helpful. If it still reproduces, sharing the training command, dataset setup details (especially around rectangular batching), and logs from the warning would help narrow down what’s left.

Thanks 🙏 for reporting this, and please share any further feedback if anything still seems off.
